src/Features/Core/Portable/AddImport/AddImportSymbolTreeInfoService.cs

```diff
@@ -0,0 +1,178 @@
+                }
+            }
+        }
+
+        internal static async Task<ImmutableArray<SymbolAndProjectId>> TryFindSourceSymbolsInCurrentProcessAsync(
+            Project project, SymbolFilter filter, SearchQuery searchQuery,
+            ConcurrentDictionary<Project, AsyncLazy<IAssemblySymbol>> projectToAssembly,
+            CancellationToken cancellationToken)
+        {
+            var service = project.Solution.Workspace.Services.GetService<ISymbolTreeInfoCacheService>();
+            var info = await service.TryGetSourceSymbolTreeInfoAsync(project, cancellationToken).ConfigureAwait(false);
+            if (info == null)
+            {
+                // Looks like there was nothing in the cache.  Return no results for now.
+                return ImmutableArray<SymbolAndProjectId>.Empty;
+            }
+
+            // Don't create the assembly until it is actually needed by the SymbolTreeInfo.FindAsync
+            // code.  Creating the assembly can be costly and we want to avoid it until it is actually
+            // needed.
+            var lazyAssembly = projectToAssembly.GetOrAdd(project, CreateLazyAssembly);
+
+            var declarations = await info.FindAsync(
+                searchQuery, lazyAssembly, project.Id,
+                filter, cancellationToken).ConfigureAwait(false);
+
+            return declarations;
+        }
+
+        private static async Task<ImmutableArray<SymbolAndProjectId>> TryFindSourceSymbolsInRemoteProcessAsync(
+            RemoteHostClient.Session session, Project project, SymbolFilter filter,
+            SearchQuery searchQuery, CancellationToken cancellationToken)
+        {
+            var array = await session.InvokeAsync<SerializableSymbolAndProjectId[]>(
+                nameof(IRemoteSymbolTreeInfoCacheService.TryFindSourceSymbolsAsync),
+                new object[] { project.Id, filter, searchQuery.Name, searchQuery.Kind }).ConfigureAwait(false);
+
+            return await ConvertAsync(project.Solution, array, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<ImmutableArray<SymbolAndProjectId>> ConvertAsync(
+            Solution solution, SerializableSymbolAndProjectId[] array, CancellationToken cancellationToken)
+        {
+            if (array == null)
+            {
+                return ImmutableArray<SymbolAndProjectId>.Empty;
+            }
+
+            var result = ArrayBuilder<SymbolAndProjectId>.GetInstance();
+
+            foreach (var value in array)
+            {
+                // Don't force getting the symbol on this end.  We don't want to perform
+                // an expensive computation locally as that may slow down Add-Import greatly.
+                var symbolAndProjectId = await value.TryRehydrateAsync(
+                    solution, forceCompilation: false, cancellationToken: cancellationToken).ConfigureAwait(false);
+
+                result.AddIfNotNull(symbolAndProjectId);
```

Comment:
I don't understand.  Does this mean we're just dropping results on the floor if this process doesn't have them?